File: classifier/data/raw/ai/samples/project_euler_java/Problem014.java
// Project Euler Problem 14: starting number below one million with the longest Collatz chain.
public class Problem014 {
    private static final int LIMIT = 1_000_000;

    public static void main(String[] args) {
        int[] lengths = new int[5_000_000];
        lengths[1] = 1;
        int bestStart = 1, bestLength = 1;
        for (int start = 2; start < LIMIT; start++) {
            long value = start;
            int steps = 0;
            while (value >= lengths.length || lengths[(int) value] == 0) {
                value = (value & 1) == 0 ? value / 2 : 3 * value + 1;
                steps++;
            }
            int length = steps + lengths[(int) value];
            if (length > bestLength) {
                bestLength = length;
                bestStart = start;
            }
            value = start;
            while (value < lengths.length && lengths[(int) value] == 0) {
                lengths[(int) value] = length--;
                value = (value & 1) == 0 ? value / 2 : 3 * value + 1;
            }
        }
        System.out.println(bestStart);
    }
}

File: classifier/data/raw/ai/samples/project_euler_zig/problem_014.zig
// Project Euler 014: starting number below one million with the longest Collatz chain.
const std = @import("std");

fn print(answer: usize) !void {
    var buffer: [32]u8 = undefined;
    const text = try std.fmt.bufPrint(&buffer, "{d}\n", .{answer});
    _ = std.c.write(1, text.ptr, text.len);
}

pub fn main() !void {
    var lengths = [_]u32{0} ** 1_000_001;
    lengths[1] = 1;
    var best_start: usize = 1;
    var best_length: u32 = 1;

    for (1..1_000_000) |start| {
        var value: u64 = start;
        var chain: [600]u64 = undefined;
        var used: usize = 0;
        while (value > 1 and (value > 1_000_000 or lengths[@intCast(value)] == 0)) {
            chain[used] = value;
            used += 1;
            value = if (value % 2 == 0) value / 2 else 3 * value + 1;
        }
        var length: u32 = if (value <= 1_000_000) lengths[@intCast(value)] else 1;
        while (used > 0) {
            used -= 1;
            length += 1;
            if (chain[used] <= 1_000_000) lengths[@intCast(chain[used])] = length;
        }
        if (lengths[start] > best_length) {
            best_length = lengths[start];
            best_start = start;
        }
    }
    try print(best_start);
}
